hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -594,25 +598,28 @@ void GlobalOptimizer::run() {
         return std::vector<NodeRef>{units.begin(), units.end()};
     }();
 
+    const auto passes__ = rt::getenv("HILTI_OPTIMIZER_PASSES");
+    const auto passes_ =
+        passes__ ? std::optional(util::split(*passes__, ":")) : std::optional<std::vector<std::string>>();
+    const auto passes = passes_ ? std::optional(std::unordered_set<std::string>(passes_->begin(), passes_->end())) :
+                                  std::optional<std::unordered_set<std::string>>();
+
+    const std::unordered_map<std::string, std::function<std::unique_ptr<OptimizerVisitor>()>> creators =
+        {{"functions", []() { return std::make_unique<FunctionVisitor>(); }},
+         {"types", []() { return std::make_unique<TypeVisitor>(); }}};
+
```

Comment:
This block I find quite hard to read, including the `if ( passes )` section further below. Maybe start with a default `std::vector<std::unique_ptr<OptimizerVisitor>> vs;` that then gets updated if `HILTI_OPTIMIZER_PASSES` is set?

Reply:
Yes, I agree this is pretty verbose. I needed to introduce the additional variables `passes__` and `passes_` since `std::optional` doesn't provide monadic bind (e.g., a `and_then` which allows applying a function over `T` on an `optional<T>`). In addition to unpacking the env var these lines also perform deduplication which could be important.

I simplified the code slightly by ensuring that `passes` is always set, either by the env var, or by overriding with a default set. This is still not trivial as we want to execute the `feature_requirements` visitor differently than other visitors while still allowing to toggle it. Let me know what you think.